fsthttp/request.go

```diff
@@ -402,12 +401,25 @@ func (req *Request) constructABIRequest() error {
 	return nil
 }
 
-func (req *Request) sendABIRequestAsyncStreaming(backend string) (*fastly.PendingRequest, *fastly.HTTPBody, error) {
+func (req *Request) sendABIRequestAsync(backend string) (*fastly.PendingRequest, *fastly.HTTPBody, error) {
 	if req.sent {
 		return nil, nil, fmt.Errorf("request already sent")
 	}
 
-	abiPending, err := req.abi.req.SendAsyncStreaming(req.abi.body, backend)
+	// When the request's ManualFramingMode is false, SendAsyncStreaming
+	// streams the request body to the backend using "Transfer-Encoding:
+	// chunked".  This is the default behavior we want for requests with
+	// a body.  For requests without a body, we want to avoid
+	// unnecessary chunked encoding, and have observed servers that
+	// error when seeing it in certain contexts.  Calling SendAsync
+	// instead causes the entire body to be buffered (in this case, zero
+	// bytes) and "Content-Length: 0" to be sent instead.
+	sendFn := req.abi.req.SendAsyncStreaming
+	if req.Body == nil {
+		sendFn = req.abi.req.SendAsync
+	}
```

Comment:
Many `io.Reader` types in the standard library expose a `Len() int` method when constructed from an in-memory buffer (`strings.Reader, `strings.Builder`, `bytes.Buffer` and `bytes.Reader` are all examples).  One possible future enhancement: do a type cast here and also use `SendAsync` when the actual request body has a known length.

Reply:
That sounds like a good plan. People tend to be surprised when chunked encoding shows up, even though it's "allowed" and a body is expected (unlike the issue giving rise to this change). We try throughout xqd and the Rust SDK to give the body a fixed length when possible to get `content-length` applied.

Reply:
https://github.com/fastly/compute-sdk-go/pull/44 implements this